[ABM-Edge-2] Implement the inventory service

Shabirmean opened the pull request:
### Depends on:
- Please review _(or make sure **merged** status of)_ the following PR before starting on this: 
   - #1 

### What's this?
- Implementation of the `inventory` service APIs
- Consists of updates to the configuration setup


### Architecture
![image](https://user-images.githubusercontent.com/7249208/131765166-afd8bf87-a3a1-4b0c-b880-2757d9536dfa.png)

***Notes on the Architecture:***
- The initial MVP will implement the data-store in memory. We will work on the expanding the app with an external storage option after the MVP
- The app exposes a _/switch_ functionality to switch between store types
- More details at: [go/abm-edge-app](http://go/abm-edge-app)

askmeegs commented on .gitignore:
```diff
@@ -2,5 +2,5 @@
 dev.yaml
-.mvn
 .DS_Store
+.mvn
```
We already chatted about this, but this should not be gitignored, I think this directory needs to be pushed upstream!

Shabirmean replied:
Fixed in https://github.com/GoogleCloudPlatform/anthos-edge-usecases/pull/2/commits/9e1f980cda2340019e2e2ef67ad52523e0641335 by merging the previous branch

askmeegs commented on point-of-sale-app/inventory/src/main/java/com/google/abmedge/inventory/InventoryController.java:
```diff
@@ -50,13 +94,92 @@ public String readiness() {
+    return new ResponseEntity<>(jsonString, HttpStatus.OK);
+  }
+
+  @PostMapping(value = "/switch/{type}")
```
Could you add a comment here, for what `/switch` does? what does it mean to change the active inventory type?

Shabirmean replied:
Added comment in https://github.com/GoogleCloudPlatform/anthos-edge-usecases/pull/2/commits/ac4ed2a65d0081ed1cbd476ca3969cb6cac4b344

askmeegs commented on point-of-sale-app/kubernetes-manifests/inventory-configmap.yaml:
```diff
@@ -0,0 +1,47 @@
+  name: inventory-configmap
+data:
+  CONNECTOR: "IN_MEMORY"
+  ACTIVE_ITEM_TYPE: "mcdonalds"
```
I recommend replacing the two well-known brand names - as well as the well known product names - with mock company/product names. Also why does the inventory span two different brand names?

Shabirmean replied:
Removed references in https://github.com/GoogleCloudPlatform/anthos-edge-usecases/pull/2/commits/945912a1e0b704603502a37fb783e622f62d6454. The idea is to be able to switch between different store-types/enterprise types without having to restart or do any major changes. These two types are here as examples

askmeegs replied:
is that switching use case something that was specifically requested? I'm trying to think of why this would be needed. wouldn't a single physical store (ie. one instance of inventory service) correspond to one store? not two?

Shabirmean replied:
Yes it was requested by @mike-ensor, as an easy way to use same deployment between multiple demos without having to restart the application. 

You are correct! This feature is not reflective of what will be out there in real; but a hidden API to enable quick switching between different kinds of store types using the same deployment without re-deploying multiple times. It's solely for convenience purpose. The API is not exposed in the UI. The idea is to have one system on multiple ABM clusters on emulated baremetal in GCE and continue to use it for multiple demos with different store types.

askmeegs commented on point-of-sale-app/inventory/src/main/java/com/google/abmedge/inventory/InventoryController.java:
```diff
@@ -50,13 +94,92 @@ public String readiness() {
+    });
+  }
+
+  private void initConnectorType() {
```
can you add a comment here on the purpose of the inventory connector?

Shabirmean replied:
Added comment here: https://github.com/GoogleCloudPlatform/anthos-edge-usecases/pull/2/commits/5f43a1ec8a1073f2e2bce71fa6298b2be8574388

askmeegs commented on point-of-sale-app/inventory/src/main/java/com/google/abmedge/inventory/InventoryController.java:
```diff
@@ -50,13 +94,92 @@ public String readiness() {
+          + "thus defaulting to: %s", ACTIVE_TYPE_ENV_VAR, ALL_ITEMS));
+      activeItemsType = ALL_ITEMS;
+    }
+    LOGGER.info(String.format("Active items type is: %s", activeItemsType));
```
what are "active" items?

Shabirmean replied:
Added a one line comment in https://github.com/GoogleCloudPlatform/anthos-edge-usecases/pull/2/commits/03fc0b4ed6c19565fe0a599a743e709052fbc1be but this is also explained in the comment added in https://github.com/GoogleCloudPlatform/anthos-edge-usecases/commit/945912a1e0b704603502a37fb783e622f62d6454

askmeegs commented on point-of-sale-app/inventory/src/main/java/com/google/abmedge/inventory/dto/Item.java:
```diff
@@ -0,0 +1,135 @@
+  private long quantity;
+  private List<String> labels;
+
+  public Item() { }
```
I'm not a java expert, but why do you have two different constructors for `Item`?

Shabirmean replied:
I only had the second one _(the one taking in 2 arguments)_ because I thought an item should always have an id and a type. Then I had to add the first one _(empty default constructor)_ because it is required by the Gson library when an object via the wire is to be deserialized and to serialize before going in the wire. Gson expects an empty constructer. 

However now looking at the code, the second constructor seems to be used only inside the same class. So I removed it and just kept the other one in 272902463e20a19857a2a27448d40abae0139c27

askmeegs commented:
✅ All pods start up, and I can call the inventory service to `GET /items`. 
<img width="589" alt="Screen Shot 2021-09-08 at 11 26 22 AM" src="https://user-images.githubusercontent.com/3137106/132539704-5fc0fcd1-ebf5-44b1-90e3-08d5c2279770.png">

Shabirmean commented:
> ✅ All pods start up, and I can call the inventory service to `GET /items`.
> <img alt="Screen Shot 2021-09-08 at 11 26 22 AM" width="589" src="https://user-images.githubusercontent.com/3137106/132539704-5fc0fcd1-ebf5-44b1-90e3-08d5c2279770.png">

@askmeegs Thank you for testing it end to end. I have addressed all your comments

puneith commented on point-of-sale-app/inventory/src/main/java/com/google/abmedge/inventory/dao/InMemoryStoreConnector.java:
```diff
@@ -0,0 +1,130 @@
+
+  @Override
+  public List<Item> getAll() {
+    return idToItemsMap.values()
```
Are these items really big that we can't do toArray() directly without invoking stream(..) API? The Java streams are a great abstraction for MR like operations but over here this looks bit overkill.

Shabirmean replied:
I am sure I understand. The `toArray` method against what object? Only reason I am stream processing each item and collecting into a list is because I need to take a deep-copy of the item before returning to ensure that mutation of it outside this API doesn't effect the underlying map.

puneith commented on point-of-sale-app/inventory/src/main/java/com/google/abmedge/inventory/dao/InMemoryStoreConnector.java:
```diff
@@ -0,0 +1,130 @@
+    if (itemId == null || StringUtils.isEmpty(itemType)) {
+      LOGGER.error("Cannot insert item. Item must have both 'id' and 'type'. "
+          + "Passed in item is missing one or both of these attributes");
+      return false;
```
Throwing Exceptions is a better way than returning a boolean flag.

Shabirmean replied:
Replaced boolean return with exception in 0f9e7029c50f6f46bed275962f5687a345e11a4d

puneith commented on point-of-sale-app/inventory/src/main/java/com/google/abmedge/inventory/dto/Item.java:
```diff
@@ -0,0 +1,133 @@
+
+  @Override
+  public String toString() {
+    return "Item{" +
```
More of a nit and stylistic as well. But a String.join (and other family of methods StringJoiner etc) is cleaner :-)

Shabirmean replied:
Refactored to use String joiner in 88abc05ee21dac979a85e1b13f6772463fb9b4bc

puneith commented on point-of-sale-app/inventory/src/main/java/com/google/abmedge/inventory/dto/PurchaseItem.java:
```diff
@@ -0,0 +1,38 @@
+  private UUID itemId;
+  private long itemCount;
+
+  public UUID getItemId() {
```
Why not have a constructor and avoid setter methods. PurchaseItem can change UUID and itemCount? Also the mutability of PurchaseItem is a question which can direct you constructor vs setter method way.

Shabirmean replied:
There is no place an object of this class is explicitly created by our code. The only place it is used in is the function argument to the [`update` API of the inventory controller](https://github.com/GoogleCloudPlatform/anthos-edge-usecases/blob/add-inventory-apis/point-of-sale-app/inventory/src/main/java/com/google/abmedge/inventory/InventoryController.java#L134). So this class is used by Springboot to deserialize a json request body into a java object when an API request arrives at /update. The serializer requires the default constructor and getter/setters.

puneith reviewed (approved):
lgtm
